### PR TITLE
feat(canary): add traffic management support

### PR DESCRIPTION
**Pull request by mochacat**
## Description

This adds traffic management as a top-level block of the deployment yaml file.

## Motivation and Context

TDD: https://armory.slab.com/posts/tdd-11-smi-traffic-management-bt031ao0#hl0mb-cli-dsl
JIRA: https://armory.atlassian.net/browse/YODL-318

## How has this been tested? How can a reviewer test these changes?

The change to the canary happy path file verifies that the the traffic management yaml gets translated correctly in the pipeline request.

# Have your performed the following tests?

Please confirm you have done the following tests to ensure your pull request is ready to be merged..

- [ ] If this change modifies the deployment flow have triggered a deployment using this branch (i.e. `build/dist/darwin_amd64/armory deploy start -f <path to file>`)?
- [ ] If this change modifes the login flow have you tried logging in and out with this branch (i.e. `build/dist/darwin_amd64/armory login`) and verified the credentials work?
- [ x ] Have you verified the specific change made in this PR?

### CODE REVIEW

File: pkg/deploy/service.go [hunk 1]
```diff
@@ -290,6 +295,40 @@ func buildStrategy(modelStrategy model.OrchestrationConfig, strategyName string)
...
+	}
+	var tms []de.KubernetesV2TrafficManagement
+	for _, tm := range *mo.TrafficManagement {
+		for _, t := range tm.Targets {
+			if t == currentTarget && tm.SMI != nil {
+				var configErrors []string
+				if tm.SMI.RootServiceName == nil {
+					configErrors = append(configErrors,"rootServiceName required in smi")
+				}
+				if tm.SMI.CanaryServiceName == nil {
```
Comment (danielpeach) [inline pkg/deploy/service.go]: Both `canaryServiceName` and `trafficSplitName` are optional: https://armory.slab.com/posts/tdd-11-smi-traffic-management-bt031ao0#he5vp-api-contract

File: pkg/deploy/testdata/sadPathDeploymentFileTrafficManagement.yaml [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    strategy: strategy2
+    constraints:
+      dependsOn: ["test1"]
+      beforeDeployment:
+        - pause:
+            untilApproved: true
+trafficManagement:
+  - targets: ["test1"]
+    smi:
+      rootServiceName: "qa-service-root"
```
Comment (danielpeach) [inline pkg/deploy/testdata/sadPathDeploymentFileTrafficManagement.yaml]: `rootServiceName` is the only one that's required - would just swap in `canaryServiceName` here.

**Review by danielpeach — COMMENTED** [review]
Looks great - just a few changes on the validation (canaryServiceName and trafficSplitName are optional).

**Comment by mochacat** [comment]
@danielpeach  add traffic management to the template:

`armory-cli template kubernetes canary -f traffic` adds this top level block:

```
trafficManagement:
    - # Specify a list of target names where the traffic management should occur. They should be defined in the top level targets block.
      targets: []
      smi:
        rootServiceName: "" # Name of the root service for deployment. The root service is required and must exist in your target environment at the time of deployment
        trafficSplitName: "" # Optional name of the service serving the new version. By default, "<rootServiceName>-canary".
        canaryServiceName: "" # Optional name of the auto-generated trafficSplit custom resource. By default "<rootServiceName>"
```

File: cmd/template/canary.go [hunk 3]
```diff
@@ -73,10 +73,24 @@ func canary(cmd *cobra.Command, options *templateCanaryOptions, args []string) e
...
-		case feature == "automated":
+		case "traffic":
+			trafficNode, trafficValuesNode := util.BuildSequenceNode("trafficManagement", "test1")
+			trafficItemNode := &yaml.Node{Kind: yaml.MappingNode, Tag: "!!map"}
+
+			targetSliceNode, targetSliceValuesNode := util.BuildSequenceNode("targets", "Specify a list of target names where the traffic management should occur. They should be defined in the top level targets block.")
+			trafficItemNode.Content = append(trafficItemNode.Content, targetSliceNode, targetSliceValuesNode)
+
+			smiNode, smiValuesNode := util.BuildMapNode("smi", "")
+			smiValuesNode.Content = append(smiValuesNode.Content, util.BuildStringNode("rootServiceName", "", "Name of the root service for deployment. The root service is required and must exist in your target environment at the time of deployment")...)
```
Comment (danielpeach) [inline cmd/template/canary.go]:
```suggestion
			smiValuesNode.Content = append(smiValuesNode.Content, util.BuildStringNode("rootServiceName", "", "Name of the root service for deployment. The root service is required and must exist in your target environment at the time of deployment.")...)
```

File: cmd/template/canary.go [hunk 4]
```diff
@@ -73,10 +73,24 @@ func canary(cmd *cobra.Command, options *templateCanaryOptions, args []string) e
...
+			smiValuesNode.Content = append(smiValuesNode.Content, util.BuildStringNode("trafficSplitName", "", "Optional name of the service serving the new version. By default, \"<rootServiceName>-canary\".")...)
+			smiValuesNode.Content = append(smiValuesNode.Content, util.BuildStringNode("canaryServiceName", "", "Optional name of the auto-generated trafficSplit custom resource. By default \"<rootServiceName>\"")...)
```
Comment (danielpeach) [inline cmd/template/canary.go]:
```suggestion
			smiValuesNode.Content = append(smiValuesNode.Content, util.BuildStringNode("canaryServiceName", "", "Optional name of the auto-generated trafficSplit custom resource. By default, \"<rootServiceName>\".")...)
```

**Review by danielpeach — DISMISSED** [review]

File: pkg/deploy/testdata/happyPathDeployEngineRequest.json [hunk 5]
```diff
@@ -47,6 +47,15 @@
       },
       "strategy": {
         "canary": {
+          "trafficManagement": [
```
Comment (danielpeach) [inline pkg/deploy/testdata/happyPathDeployEngineRequest.json]: The format here looks wrong... but I'm not sure this is being used anywhere.

**Review by danielpeach — DISMISSED** [review]

File: pkg/deploy/testdata/happyPathDeployEngineRequest.json (hunk 5 above)
Comment (mochacat) [reply pkg/deploy/testdata/happyPathDeployEngineRequest.json]: Should be updated for the DE multi-deployment format now

**Review by danielpeach — APPROVED** [review]

**Pull request merged**